lib/jnpr/junos/factcache.py

```diff
@@ -0,0 +1,280 @@
+                else:
+                    string = current
+        return '{' + string + '}'
+
+    def _refresh(self,
+                 exception_on_failure=False,
+                 warnings_on_failure=False,
+                 keys=None):
+        """
+        Empty the cache to force a refresh of one or more facts.
+
+        Empties the fact gathering cache for all keys (if keys == None) or a
+        set of keys. This causes the fact to be gathered and cached upon next
+        access. If either eception_on_failure or warnings_on_failure is true,
+        then all facts are accessed by getting the string representation of the
+        facts. This causes all facts to immediately be populated so that any
+        exceptions or warnings are generated during the call to _refresh().
+
+        :param exception_on_failure: A boolean which indicates if an exception
+          should be raised upon a failure gathering facts.
+
+        :param warnings_on_failure: A boolean which indicates if an warning
+          should be logged upon a failure gathering facts.
+
+        :param keys: A single key as a string, or an iterable of keys as a list
+          or tuple. The specified keys are emptied from the cache. If None, all
+          keys are emptied from the cache.
+
+        :raises RuntimeError:
+            When keys contains an unknown fact.
+        """
+        refresh_keys = None
+        if keys is not None:
+            if isinstance('str', type(keys)):
+                refresh_keys = (keys,)
+            else:
+                refresh_keys = keys
+        if refresh_keys is not None:
+            for key in refresh_keys:
+                if key in self._cache:
+                    del self._cache[key]
+                else:
+                    raise RuntimeError('The %s fact can not be refreshed. %s '
+                                       'is not a known fact.' % (key, key))
+        else:
+            self._cache = dict()
+        if exception_on_failure or warnings_on_failure:
+            self._exception_on_failure = exception_on_failure
+            self._warnings_on_failure = warnings_on_failure
+            try:
+                _ = str(self._device.facts)
+            except Exception:
+                if exception_on_failure:
+                    raise
+            finally:
+                if warnings_on_failure and self._should_warn:
+                    warnings.warn('Facts gathering is incomplete. '
+                                  'To know the reason call '
+                                  '"dev.facts_refresh('
+                                  'exception_on_failure=True)"',
+                                  RuntimeWarning)
+                self._exception_on_failure = False
+                self._warnings_on_failure = False
+                self._should_warn = False
+
+# Make this class look like a regular dict.
+# Ensures isinstance() and issubclass() behave the same as for a dict.
```

Comment:
Yes. This is what I was trying to workaround with using the ABCMeta metaclass. Unfortunately, what I had implemented was the inverse of what was intended (an instance of a dict was an instance of _FactCache, but not vice-versa.)